### ISSUE TITLE
[MU4 Issue] Mixer - Auto-resizing issues when docked with tabs

### ISSUE THREAD

**Issue by lengthwave** [musescore/MuseScore#11335]
**Describe the bug**
Mixer window does not automatically resize height properly when tabbed due to other features also docked (ex: timeline)

**To Reproduce**
1. Open Mixer and dock
2. Open Timeline and dock
3. From three dot menu in Mixer tab select "View > Volume"
4. From three dot menu in Mixer tab desect select "View > Volume"
5. See error in window height auto-resize

**Expected behavior**
Mixer window should automatically resize when tabbed in the same way as when mixer is the only feature docked.

**Screenshots**
<img width="1919" alt="Screen Shot 2022-04-25 at 1 36 57 PM" src="https://user-images.githubusercontent.com/30370609/165073061-b80eec98-3523-4dd5-ab99-de14f5e9d7b8.png">

**Platform information**
 OS: macOS 12.2, Arch.: x86_64
MuseScore version (64-bit): 4.0.0-2214419458
revision: github-musescore-musescore-df96a36

**Comment by lengthwave** [comment]
@Tantacrul - FYI

**Comment by cbjeukendrup** [comment]
I'd like to add that I've seen several reports of people getting really confused by this: they thought they couldn't choose Muse Sounds anymore, but the reason was that they were looking in the FX dropdown instead of in the Sound dropdown, because the Sounds dropdown was completely hidden behind the title bar.

**Comment by Tantacrul** [comment]
I'm adding this to our design tasks because there are a few fixes that need doing with the bottom docking bar, including fixing this highly annoying issue.

**Comment by daijingz** [comment]
@cbjeukendrup I want to do this issue, but I am not sure whether it is too difficult for me.

**Comment by cbjeukendrup** [comment]
@daijingz Feel free to give it a try, but it might be a little bit tricky. I'll give you one hint: start inspecting the code in MixerPanel.qml, and look what happens when the implicitHeight of the panel changes (because that is the trigger to resize the panel). From there, there seems to be a relatively straightforward path to the solution, but it does require some knowledge of the KDDockWidgets library and about how properties of QObjects work, and that's what makes it tricky.

If you really need some more hints at some point, feel free to ask in #development on Discord (that's highly preferred over private messages).

**Comment by daijingz** [comment]
Before adding columns: (With timeline)
![UEME8{3$WEB7ZQ3GSNL9LKM](https://github.com/musescore/MuseScore/assets/61481741/c420d190-a7a4-482d-a1c2-0f804d0fd0b8)
After adding columns: (Without timeline)
![LX_RUS$}A451Y%503O8(TIK](https://github.com/musescore/MuseScore/assets/61481741/59b4e315-6d85-402b-a03f-a481328d56c4)

**Comment by daijingz** [comment]
@lengthwave @cbjeukendrup I guess your meaning is when both the timeline and mixer are docked, the mixer does not work well in its resizing function. (When only the mixer is docked, this problem does not occur).
I want to see where the code of the mixer and timeline page is defined.

**Comment by daijingz** [comment]
I suspect the problem occurs at contentColumn.completed

**Comment by cbjeukendrup** [comment]
@daijingz `onImplicitHeightChanged` is indeed the correct starting point. The next logical thing to do is to look what `resizeRequested` is: it is a signal of MixerPanel. So if it's a signal, then there must be some handler for it somewhere, because otherwise nothing would happen. So you search for `onResizeRequested`. You should find this in some other file.

But it seems you don't understand yet what's actually the issue here. The issue is that calling `resizeRequested` does not produce the expected result if there are tabs instead of just a title bar. The height of the panel becomes the `implicitHeight` of the MixerPanel, but it should become the `implicitHeight` of the MixerPanel _PLUS_ the height of the tabs, otherwise the top of the MixerPanel is not visible. See the illustrations below:
| Result after `resizeRequested` with just a title bar (correct) | Current result after `resizeRequested` with tabs (incorrect) | Desired result after `resizeRequested` with tabs |
|--------|--------|--------|
| <img width="689" alt="Scherm­afbeelding 2024-03-17 om 17 51 21" src="https://github.com/musescore/MuseScore/assets/48658420/d73ecebd-a797-471b-9e4d-a669dbbf4dcc"> | <img width="689" alt="Scherm­afbeelding 2024-03-17 om 17 36 05" src="https://github.com/musescore/MuseScore/assets/48658420/d5532ea6-351a-4328-bdc4-2825b8374ab3"> | <img width="689" alt="Scherm­afbeelding 2024-03-17 om 17 46 45" src="https://github.com/musescore/MuseScore/assets/48658420/a4835c03-b6ba-40bb-8e28-d2e3b62cb86b"> | 

(Click the images to view them larger)

**Comment by daijingz** [comment]
Okay, so you want to add the height of the title bar with the `implicitHeight`

**Comment by cbjeukendrup** [comment]
Yes, but not at the place where `resizeRequested` is being called. The MixerPanel has namely no way to know the height of the tab bar, nor whether there is currently a tab bar of a title bar.

Your goal is now to figure out where the height of the title bar gets added, and then make sure that the tab bar is handled at that place too. To find the relevant code, you'll have to look which functions are called when `resizeRequested` is called.

**Comment by daijingz** [comment]
> @daijingz `onImplicitHeightChanged` is indeed the correct starting point. The next logical thing to do is to look what `resizeRequested` is: it is a signal of MixerPanel. So if it's a signal, then there must be some handler for it somewhere, because otherwise nothing would happen. So you search for `onResizeRequested`. You should find this in some other file.
> 
> But it seems you don't understand yet what's actually the issue here. The issue is that calling `resizeRequested` does not produce the expected result if there are tabs instead of just a title bar. The height of the panel becomes [the `implicitHeight` of the MixerPanel], but it should become [the `implicitHeight` of the MixerPanel] _PLUS_ [the height of the tabs], otherwise the top of the MixerPanel is not visible. See the illustrations below:
> 
> Result after `resizeRequested` with just a title bar (correct)	Current result after `resizeRequested` with tabs (incorrect)	Desired result after `resizeRequested` with tabs
> <img alt="Scherm­afbeelding 2024-03-17 om 17 51 21" width="689" src="https://private-user-images.githubusercontent.com/48658420/313490451-d73ecebd-a797-471b-9e4d-a669dbbf4dcc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTY0NjMsIm5iZiI6MTcxMDY5NjE2MywicGF0aCI6Ii80ODY1ODQyMC8zMTM0OTA0NTEtZDczZWNlYmQtYTc5Ny00NzFiLTllNGQtYTY2OWRiYmY0ZGNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE3VDE3MjI0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiMWZjZTgwOGNlMzVkNGExZDg5NDgyYjAzM2E1NWExZGU4N2Q1NWRjYjUyYzg2YmM1NmNjMzNjYjM5MWFhMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rauy5amdyuPCTvwC8EcNYOUnDZEwzBf9bnzh_p79pe8">	<img alt="Scherm­afbeelding 2024-03-17 om 17 36 05" width="689" src="https://private-user-images.githubusercontent.com/48658420/313489823-d5532ea6-351a-4328-bdc4-2825b8374ab3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTY0NjMsIm5iZiI6MTcxMDY5NjE2MywicGF0aCI6Ii80ODY1ODQyMC8zMTM0ODk4MjMtZDU1MzJlYTYtMzUxYS00MzI4LWJkYzQtMjgyNWI4Mzc0YWIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE3VDE3MjI0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiMzEzZTExMGRhZjVkNmZjMTA3MDVlMmEyMjNiZTgzZjZmYTE2N2VjYzUyZGFhMmE1MDIyM2IzMjE2OTJjZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QtqZS3zxeIqgnXI2nrnpkNKB4lAFMo67Vpjx6ZxENY4">	<img alt="Scherm­afbeelding 2024-03-17 om 17 46 45" width="689" src="https://private-user-images.githubusercontent.com/48658420/313490266-a4835c03-b6ba-40bb-8e28-d2e3b62cb86b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTY0NjMsIm5iZiI6MTcxMDY5NjE2MywicGF0aCI6Ii80ODY1ODQyMC8zMTM0OTAyNjYtYTQ4MzVjMDMtYjZiYS00MGJiLThlMjgtZDJlM2I2MmNiODZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE3VDE3MjI0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5NzQzNDk4NTZiZTUxYzMzMGEyMGFjZWFiYzNjZTM2Y2EwZTU1OTJiOGUzOTZlNDY4MjkzNTM1NjBiYzA5NjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.k-Z1Am9hpjRzn5y2M57ndOY2qwm-mhFOtsxOBGhnmCs">
> (Click the images to view them larger)

![N8I31@S_7U8K{CJK39V{()I](https://github.com/musescore/MuseScore/assets/61481741/8708e48c-3dcb-494f-89a8-9c78ca397af7)

**Comment by cbjeukendrup** [comment]
Yes, that's the `onResizeRequested` handler. So the next logical thing to do, is to look what this `mixerPanel.resize` function does / where it is defined. 

One tip: this function is defined in C++, so you'll have to do some extra steps to find it. 

You don't need to post every intermediate step here, but once you've found the function, feel free to post an update again, including _how_ you found it.

By the way, I really recommend you invest some time in setting up Qt Creator. It really makes life easier, because then you get syntax highlighting, auto-complete, and Ctrl+click for "jump to definition". https://github.com/musescore/MuseScore/wiki/Compile-in-Qt-Creator

**Comment by daijingz** [comment]
> Yes, that's the `onResizeRequested` handler. So the next logical thing to do, is to look what this `mixerPanel.resize` function does / where it is defined.
> 
> One tip: this function is defined in C++, so you'll have to do some extra steps to find it.
> 
> You don't need to post every intermediate step here, but once you've found the function, feel free to post an update again, including _how_ you found it.
> 
> By the way, I really recommend you invest some time in setting up Qt Creator. It really makes life easier, because then you get syntax highlighting, auto-complete, and Ctrl+click for "jump to definition". https://github.com/musescore/MuseScore/wiki/Compile-in-Qt-Creator

I am just going to update most information, and you do not need to take care of everything. Don't worry.

**Comment by daijingz** [comment]
I think this is the corresponding code:
![MBT0NJ8}3 %W1ILOY3S`RWJ](https://github.com/musescore/MuseScore/assets/61481741/1123c17c-f14f-42b1-bcbd-fd6aa9b29823)

**Comment by daijingz** [comment]
To change the height, this is the height update code section:
```
    const QQuickItem* visualItem = frame->visualItem();
    int extraHeight = visualItem ? visualItem->property("nonContentsHeight").toInt() : 0;
    height += extraHeight;
```

**Comment by cbjeukendrup** [comment]
For the record, see my reply on Discord explaining a bit about the property system in Qt (https://doc.qt.io/qt-6/properties.html).

The conclusion was that searching for the property name, i.e. `nonContentsHeight` will lead you to the place where its value comes from. (Hint: it's in QML.)

**Comment by bkunda** [comment]
@daijingz @cbjeukendrup we've actually updated the design spec for this issue (only last week!), and @jessjwilliamson is going to raise it as a new task fairly shortly. Not sure how that will impact the work that's been happening here in the past few days, but it might be prudent to wait until the new design spec is raised before spending further time on this.

**Comment by cbjeukendrup** [comment]
I think we can go ahead anyway, since it's trivial enough and if my guess about the new design is anywhere near correct we would benefit from this fix anyway.

**Comment by jessjwilliamson** [comment]
Task made https://github.com/musescore/MuseScore/issues/22016

**Comment by daijingz** [comment]
@cbjeukendrup @jessjwilliamson @bkunda I am still confused (Please don't leave me behind), about these hesitations, should we continue working on this issue or not?

**Comment by cbjeukendrup** [comment]
It turns out the issue will already be fixed by #22050, so you can switch to a different task
